### ISSUE TITLE
[Bug]: FROST KDA produces NaN/Inf with correlated BF16 keys and safe_gate=True

### ISSUE THREAD

**Issue by mdy666** [NVIDIA/cudnn-frontend#1051]
## Problem

FROST KDA returns NaN/Inf for finite BF16 inputs with strongly correlated keys, even with `safe_gate=True, gate_lower_bound=-5`. Both `batch_invariant=False` and `True` are affected.

Environment: NVIDIA B300, PyTorch 2.9.0, CUDA 13.0, cuDNN Frontend 1.29.0, CuTe DSL 4.7.0.

The repro uses only generated data, one sequence (`cu_seqlens=int32 [0,8192]`), no initial state, and two heads of dimension 128. Key norms are about 0.9, beta is 1, and the activated log gates are approximately [-0.0341, -0.0121].

## Expected / actual

An independent FP64 token-by-token recurrence remains finite (`max(abs(output)) = 0.0097991`). cuDNN produces:

| batch_invariant | Nonfinite output elements |
| --- | ---: |
| False | 676592 |
| True | 676768 |

All seven returned input/parameter gradient tensors also contain NaN/Inf. Replacing the correlated keys with independent normalized random keys keeps outputs and gradients finite; forward relative RMS error is approximately 0.43%.

## Reproducer

Save as `repro.py` and run `python3 repro.py`. It prints results for both modes and ends with an assertion failure. No data files or additional model package are required.

```python
import os
os.environ.setdefault("CUBLAS_WORKSPACE_CONFIG", ":4096:8")

import torch
import torch.nn.functional as F
from cudnn.linear_attention import kimi_delta_attention

torch.set_num_threads(1)
torch.manual_seed(42)
torch.backends.cuda.matmul.allow_tf32 = False
T, H, D = 8192, 2, 128
shape = (T, H, D)

# CPU generation makes the data independent of the GPU's SM count.
q = F.normalize(torch.randn(shape), dim=-1)
direction = F.normalize(torch.randn(1, H, D), dim=-1)
k = 0.9 * F.normalize(direction + 0.002 * torch.randn(shape), dim=-1)
v, g = 0.1 * torch.randn(shape), 0.1 * torch.randn(shape)
cpu = {name: x.bfloat16() for name, x in zip(("q", "k", "v", "g"), (q, k, v, g))}
cpu.update(beta=torch.ones(T, H, dtype=torch.bfloat16),
           a_log=torch.zeros(H), dt_bias=torch.full((H, D), -5.5))
dy = torch.randn(shape).bfloat16()
assert all(torch.isfinite(x).all() for x in cpu.values())

# Independent FP64 token-by-token recurrence, with exactly the quantized inputs.
q, k, v, g, beta, a_log, dt_bias = [x.double() for x in cpu.values()]
alpha = (-5 * torch.sigmoid(a_log.exp()[:, None] * (g + dt_bias))).exp()
state = torch.zeros(H, D, D, dtype=torch.float64)
ref = []
for t in range(T):
    state = alpha[t, :, :, None] * state
    residual = v[t] - torch.einsum("hk,hkv->hv", k[t], state)
    state = state + beta[t, :, None, None] * k[t, :, :, None] * residual[:, None, :]
    ref.append(torch.einsum("hk,hkv->hv", q[t] / D**0.5, state))
ref = torch.stack(ref)
assert torch.isfinite(ref).all()
print("reference max_abs:", ref.abs().max().item(), flush=True)

passed = True
for bi in (False, True):
    xs = {name: x.cuda().requires_grad_() for name, x in cpu.items()}
    y = kimi_delta_attention(
        **xs, cu_seqlens=torch.tensor([0, T], device="cuda", dtype=torch.int32),
        initial_state=None, output_final_state=False,
        safe_gate=True, gate_lower_bound=-5.0,
        batch_invariant=bi, plan_name="kda_frost",
    )[0]
    grads = torch.autograd.grad(y, tuple(xs.values()), dy.cuda())
    torch.cuda.synchronize()
    nonfinite = {"output": int((~torch.isfinite(y)).sum())}
    nonfinite.update({"d_" + name: int((~torch.isfinite(grad)).sum())
                      for name, grad in zip(xs, grads)})
    print("batch_invariant:", bi, "nonfinite:", nonfinite, flush=True)
    passed = passed and not any(nonfinite.values())
assert passed, "cuDNN returns NaN/Inf for finite bounded inputs; FP64 reference stays finite"
```

Possible location: the BF16 intermediate casts in the 16×16 Neumann inverse in `frost/kernel/kda_prefill_f16.py`. Correlated keys appear to amplify cancellation error and destabilize the recurrent state; this is not an input NaN or an out-of-range safe-gate value.

**Comment by mdy666** [comment]
Retested the exact CPU-seeded reproducer against clean `develop` Python sources at `b2712de2da3832e9ad607867d22b2a1429c89cdc` (native Frontend 1.29.0, cuDNN backend 9.24.0, B300 / PyTorch 2.9.0 / CUDA 13.0 / CuTe DSL 4.7.0). The failure persists with the same nonfinite output counts: 676592 for BI off and 676768 for BI on; all seven gradients also contain NaN/Inf. The FP64 recurrence remains finite.

The investigation points to precision loss in the BF16 intermediate powers of the 16×16 Neumann inverse. An experimental FP32 forward-substitution replacement makes the same long-input case finite, with about 1.32% forward relative RMS error against FP64. This is not yet a complete fix: shorter correlated-key cases still exceed the backward reference tolerance, so I am not proposing that experiment as a ready-to-merge PR.

The remaining parameter-gradient error is mostly upstream of the final gate reduction. On a 16-token correlated-key case with the experimental solve, the incoming log-gate gradient has 0.902% relative RMS error, amplified by cancellation to 24.85% in `d_a_log`. Recomputing only that final reduction in FP64 from the same incoming derivative differs from the kernel by just 0.0407%. Storing the raw gate in FP32 does not remove the discrepancy.

All these checks use synthetic data with `safe_gate=True, gate_lower_bound=-5`; no tolerance was relaxed.

**Comment by jhjpark** [comment]
I put a PR that should hopefully address this. Instead of a Neumann inverse on the whole 16x16 matrix, it does it for each 4x4 diagonal block and then does block matrix inversion for the rest. Locally, I think this addresses the repro script. Let me know if this looks good.
